MWPW-152792: Adjust CSS in Catalog card for promo treatment

mirafedas opened the pull request:
- Added a slot 'payment-details' located inside the 'heading-m' slot;
- Restricted the catalog badge max width to 140px;

Changes related to the callout text were added in [this PR](https://github.com/adobecom/milo/pull/2535), I only moved them into a separate function here. Adding this info here as looks like it was not documented anywhere: 
to add a tooltip for the callout text, place the link to the tooltip icon with the text '#icon' and add the tooltip text in the link after the hash, example: `https://main--milo --adobecom.hlx.page/drafts/rosahu/info-icon.svg#this%20is%20a%20dummy%20tooltip%20text`

Updated authoring documentation: https://main--  milo--mirafedas.hlx.live/docs/authoring/commerce/merch-card-catalog

Resolves: [MWPW-152792](https://jira.corp.adobe.com/browse/MWPW-152792)

**Test URLs:**
- Before: https://main--cc --adobecom.hlx.page/es/cc-shared/fragments/merch/products/catalog/merch-card/cc/uat/promo-test/ai-promo-test?martech=off&georouting=off
- After: https://mwpw-152792-badge-vat--milo--mirafedas.hlx.page/drafts/mirafedas/do-not-modify/catalog-card-vat-label?martech=off&georouting=off

overmyheadandbody commented on libs/blocks/merch-card/merch-card.js:
```diff
@@ -210,36 +210,43 @@ const parseContent = async (el, merchCard) => {
+    if (tagName === 'H6') {
+      if (element.firstChild.nodeType === Node.TEXT_NODE) {
+        const paymentDetails = createTag('div', { slot: 'payment-details' });
+        paymentDetails.innerHTML = element.innerHTML;
```
This might trigger some Kodiak issues, maybe a spread or recursive append would be better suited. This could also be troublesome if some event listeners have already been added to `element`. Also, could `element` be used after this piece of logic? If not, maybe it could be deleted.

Axelcureno replied:
```suggestion
        const paymentDetails = createTag('div', { slot: 'payment-details' }, element.innerHTML);
```

3ch023 reviewed (changes requested):
<img width="693" alt="image" src="https://github.com/user-attachments/assets/e3bb999e-fe4f-44db-9beb-ee563907e444">

it seems to still overlap with the drop down

Axelcureno commented on libs/blocks/merch-card/merch-card.js:
```diff
@@ -210,36 +210,43 @@ const parseContent = async (el, merchCard) => {
+        const headingM = merchCard.querySelector('h4[slot="heading-m"]');
+        headingM?.append(paymentDetails);
+      } else if (element.firstElementChild?.tagName === 'EM') {
+        const calloutContentWrapper = createTag('div');
```
This `parseContent()` function is getting bulky and convoluted, why not split some logic into separate functions for easier reading and modularization, we can start by moving the callOut logic into its separate function.

Axelcureno reviewed (approved):
Approving with 2 small suggestions which are not blockers.

Axelcureno commented on libs/features/mas/web-components/src/merch-card.css.js:
```diff
@@ -140,6 +140,7 @@ export const styles = css`
         height: fit-content;
         flex-direction: column;
         width: fit-content;
+        max-width: 140px;
```
Try adding `box-sizing: border-box;` to enforce `max-width`, otherwise it will still overlap.
![Screenshot 2024-08-05 at 11 02 47 AM](https://github.com/user-attachments/assets/e1436f05-29b6-4f28-b1c8-8a3606c84917)

Axelcureno reviewed (changes requested):
As @3ch023 stated, content is still overlapping. I have made one suggestion to fix it.

rahulgupta999 commented on libs/blocks/merch-card/merch-card.js:
```diff
@@ -248,12 +266,6 @@ const parseContent = async (el, merchCard) => {
-  if (calloutContainer.children.length > 0) {
-    const calloutSlot = createTag('div', { slot: 'callout-content' });
-    calloutSlot.appendChild(calloutContainer);
-    merchCard.appendChild(calloutSlot);
```
@rohitsahu does the order of adding the callout slot matter

mirafedas replied:
Thanks guys for bringing my attention to this, I rewrote the code, not all works as it should

Axelcureno commented:
Conflicts must be resolved please

yesil commented on libs/blocks/merch-card/merch-card.js:
```diff
@@ -154,6 +154,52 @@ const parseTwpContent = async (el, merchCard) => {
 
+const appendPaymentDetails = (element, merchCard) => {
+  if (element.firstChild.nodeType !== Node.TEXT_NODE) return;
+  const paymentDetails = createTag('div', { slot: 'payment-details' }, element.innerHTML);
```
can you use `element` instead of `element.innerHTML` ?

mirafedas replied:
I'm passing `element.innerHTML` because I want to strip off the `<h6>` tag - semantically this text is not a heading, and I think there is no need to wrap this text in an additional tag. Is there any reason why we should change it to `element`?

yesil replied:
if you know that `element` is never an element with an even listener attached, that should be ok, if not, this will cause the listeners to be removed from the element.

yesil commented on libs/blocks/merch-card/merch-card.js:
```diff
@@ -154,6 +154,52 @@ const parseTwpContent = async (el, merchCard) => {
+  const fragment = document.createDocumentFragment();
+
+  emElement.childNodes.forEach((child) => {
+    if (child.nodeType === Node.ELEMENT_NODE && child.tagName === 'A' && child.innerText.trim().toLowerCase() === '#icon') {
```
Can you describe what does ` '#icon'` in the PR description?

yesil commented on libs/features/mas/web-components/src/merch-card-collection.js:
```diff
@@ -283,7 +283,7 @@ export class MerchCardCollection extends LitElement {
                 id="sortButton"
                 size="m"
-                @change="${this.sortChanged}"
+                onchange="${this.sortChanged}"
```
why this change?  this seems wrong to me.

mirafedas replied:
This change got there by mistake, thanks for noticing!

yesil commented on libs/deps/mas/merch-card-collection.js:
```diff
@@ -128,7 +128,7 @@ import{html as a,LitElement as A}from"/libs/deps/lit-all.min.js";var m=class{con
                 id="sortButton"
                 size="m"
-                @change="${this.sortChanged}"
+                onchange="${this.sortChanged}"
```
I thought I already asked about this change? what is the reason?

mirafedas replied:
Yes, sorry, forgot to build mas. It should be good now

mirafedas commented:
@yesil According to designs `paymentDetails` is just a text and it's doubtful there will be any event listeners on it, so for the sake of DOM simplicity I will leave it as it is, but thanks for the heads-up :)

yesil commented on libs/features/mas/web-components/src/merch-card.js:
```diff
@@ -432,7 +432,9 @@ export class MerchCard extends LitElement {
                 <slot name="heading-xs"></slot>
-                <slot name="heading-m"></slot>
+                <slot name="heading-m">
+                    <slot name="payment-details"></slot>
```
A `slot` inside another `slot` doesn't seem to be allowed as per chatgpt but the specs do not mention it.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot

technically this should be equal to:


```html
<slot name="payment-details"></slot>
<slot name="heading-m"></slot>
...
</slot>
``` 

what is the reason for nesting?

mirafedas replied:
@yesil 
The merch card body has a flex gap between all elements (see screenshot).
If I place the `payment-details` slot outside the `heading-m` slot, there will be a gap between the two elements and it will not look as on the XD.

![Screenshot 2024-08-13 at 10 52 54](https://github.com/user-attachments/assets/31c3c1b3-05c9-4fb0-9509-787db40d0468)

yesil replied:
@mirafedas in this case, why it can't be a part of `heading-m` ?

mirafedas replied:
@yesil we need a way to distinguish which text goes into `heading-m` and which into `payment-details` because both texts should be on different lines and have different styles

yesil replied:
@mirafedas we need to find a way to do it without nested slots though.

mirafedas replied:
I replaced the slot with a div, now should be good 👍

3ch023 commented on .github/codeql/codeql-config.yml:
```diff
@@ -5,3 +5,4 @@ paths-ignore:
   - 'tools/floodgate/**'
   - 'tools/translation/**'
   - node_modules
+  - libs/deps/mas
```
cool

Roycethan commented:
@mirafedas Can you plz check these:
1) VAT label style is not yet fixed.having larger and non-italized font... 
per this ticket - "The VAT label style is also the same size as the price, and needs a smaller italicized text style"
https://main--cc--adobecom.hlx.page/es/cc-shared/fragments/merch/products/catalog/merch-card/cc/uat/promo-test/ai-promo-test?milolibs=mwpw-152792-badge-vat--milo--mirafedas
2) Per merged PR https://github.com/adobecom/milo/pull/2663 the call out text is moved to the bottom, but in the below example its not:
https://mwpw-152792-badge-vat--milo--mirafedas.hlx.page/drafts/mirafedas/do-not-modify/catalog-card-vat-label?martech=off&georouting=off

mirafedas commented:
@Roycethan 
1. I fixed it by changing the `{{small-tax-incl-label}}` to H6.
2. To place the promo on the bottom, authors must set the `promo bottom` parameter on the card. I added a second card with this parameter, now we have both cases covered.
